File: sorted_two_sum.rb
arr = [2, 4, 6, 8, 9, 10, 14, 16, 20]
target = 9

# arr = [2, 4, 5, 7, 8, 10]
# target = 11

# arr = [1, 4, 5, 6, 8, 9]
# target = 11


# Given a sorted array of integers and a target value, determine if there exists two integers in the array that sum up to the target value.

def sorted_two_sum(arr, target)
  left = 0
  right = arr.length - 1

  while (arr[left] + arr[right]) != target do
    if left > right
      return false
    end

    sum = arr[left] + arr[right]

    if sum > target
      right -= 1
    end

    if sum < target
      left += 1
    end
  end

  return true
end

p sorted_two_sum(arr, target)

File: sortedTwoSum.js
// Given a sorted array of integers and a target value, determine if there exists two integers in the array that sum up to the target value.

// Constraints
// Time Complexity: O(N)
// Auxiliary Space Complexity: O(1)

// let arr = [1, 5, 13, 2, 9, -6, 18, 3];
const arr = [1, 2, 5, 8, 13, 22];
const target = 36;

function sortedTwoSum(arr, target) {
  let leftPointer = 0;
  let rightPointer = arr.length - 1;

  while (leftPointer < rightPointer) {
    if (arr[leftPointer] + arr[rightPointer] === target) {
      return true;
    } else if (arr[leftPointer] + arr[rightPointer] > target) {
      rightPointer--;
    } else if (arr[leftPointer] + arr[rightPointer] < target) {
      leftPointer++;
    }
  }

  return false;
}

// function unsortedTwoSum(arr, target) {
//   let diffs = {};

//   for (let i = 0; i < arr.length; i++) {
//     let currentInt = arr[i];
//     let diffInt = target - currentInt;

//     diffs[diffInt] = 1;

//     if (diffs[currentInt]) {
//       return true;
//     }
//   }

//   return false;
// }

console.log("\nSolution:\n", sortedTwoSum(arr, target));